### PR TITLE
Adagrad max

### LINKED ISSUE #46480: Allow negative learning rates
## 🚀 Feature
Currently, optimizers throw an assertion error when negative learning rates are supplied at construction. This proposal suggests removing this restriction. 

## Motivation
Since maximization is equivalent to minimizing a negative loss function, negative learning rates are useful and make sense. For example, in GANs, the generator and discriminator can be trained adversarially by giving the discriminator a negative learning rate. This avoids having two backward passes, which improves both computational efficiency and conceptual clarity. 

## Pitch
As learning rates are typically parameterized by constants, providing a negative rate by accident is highly unlikely. I argue that making this mistake is much less likely than wanting a negative learning rate, and so the defensive assertion is better removed. 

## Alternatives
It is currently possible to set a negative LR through an ugly loop through the optimizer's `param_group`. An alternative would be to provide a cleaner way to do so, i.e., a `maximize=True` flag on optimizer construction. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @vincentqb @iramazanli

### PR DESCRIPTION
**Pull request by wukong1992**
Fixes #46480 --adagrad

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/optim/_functional.py [hunk 1]
```diff
@@ -33,8 +34,7 @@ def adagrad(params: List[Tensor],
                 raise RuntimeError("weight_decay option is not compatible with sparse gradients")
             grad = grad.add(param, alpha=weight_decay)
 
-        clr = lr / (1 + (step - 1) * lr_decay)
-
+        clr = lr / (1 + (step - 1) * lr_decay) if not maximize else -lr / (1 + (step - 1) * lr_decay)
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/optim/_functional.py:37]: Is flipping the learning rate really the right thing to do here? Isn't flipping the gradients a better thing to do here to make sure that the (non-linear) statistic computation is done correctly?

[the remark this answers is not in the document]
Comment (wukong1992) [reply torch/optim/_functional.py]: > Is flipping the learning rate really the right thing to do here? Isn't flipping the gradients a better thing to do here to make sure that the (non-linear) statistic computation is done correctly?

Sorry， I don’t understand "the (non-linear) statistic computation is done correctly" , can you explain？ I think flipping ‘grad’ and ‘lr’  both can achieve the goal，Is flipping the grad will affect other  logic outside optimizer? The ‘grad’ is a reference or a value? If is the value , I think both is ok.  I don't know if I understant right, if there is something wrong, please correct it.

**Comment by wukong1992** [comment]
I can't figure out why mypy ci error: `torch/distributed/optim/functional_adagrad.py:94:32: error: Argument "maximize" to "adagrad" has incompatible type "float"; expected "bool"  [arg-type]`, the `maximize` param is bool type. @albanD

File: torch/optim/_functional.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/optim/_functional.py]: Given that the two don't do the same thing, we need to choose one and we should have a good argument to use one over the other.
